src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Patch.cs

```diff
@@ -0,0 +1,595 @@
+                if (expectsTableArgs ? !curType.IsTable : !curType.IsRecord)
+                {
+                    errors.EnsureError(args[i], TexlStrings.ErrNeedRecord_Arg, args[i]);
+                    continue;
+                }
+
+                // Ensure that if the key in argument1 exists in the current record, their types match.
+                foreach (var typedName in curType.GetNames(DPath.Root))
+                {
+                    // As long as source doesn't contain control type we can skip the type check for control types.
+                    if (!sourceContainsControlType && typedName.Type.IsControl)
+                    {
+                        continue;
+                    }
+
+                    // If the datasource doesn't contain the supplied type.
+                    DName name = typedName.Name;
+                    if (!dataSourceType.TryGetType(name, out DType dsNameType))
+                    {
+                        dataSourceType.ReportNonExistingName(FieldNameKind.Display, errors, name, args[i]);
+                        continue;
+                    }
+
+                    DType type = typedName.Type;
+
+                    // If the type has metafield in it then it's coming from a control type like dropdown or listbox.
+                    // For example, dropdown.SelectedItems, listbox.SelectedItems. So expand the type to include property types as well.
+                    // In Document.cs!AugmentedExpandoType function, we don't add the field to the expandotype if it conflicts with any expando property.
+                    // So the type doesn't include that field. This logic tries to rectify that by adding the missing fields back.
+                    // This is only necessary if we're trying to compare it to an entity or aggregate type, as the meta field will expand to a table or record
+                    if (type.HasMetaField() && (dsNameType.IsAggregate || dsNameType.Kind == DKind.DataEntity) && type.IsAggregate)
+                    {
+                        type = ExpandMetaFieldType(type);
+                    }
+
+                    // For patching entities, we expand the type and drop entities and attachments for the purpose of comparison.
+                    if (dsNameType.Kind == DKind.DataEntity && type.Kind != DKind.DataEntity)
+                    {
+                        if (!dsNameType.TryGetExpandedEntityTypeWithoutDataSourceSpecificColumns(out _))
+                        {
+                            binding.DeclareMetadataNeeded(dsNameType);
+                        }
+                    }
+                }
+            }
+        }
+    }    
+
+    internal abstract class PatchAsyncFunctionCore : PatchAndValidateRecordFunctionBase
+    {
+        public override bool IsAsync => true;
+
+        public override bool ManipulatesCollections => true;
+
+        // !!!TODO Shouldn't this be false?
+        public override bool IsSelfContained => true;
```

Comment:
>l IsSelfContained => true; [](http://example.com/codeflow?start=27&length=26)

This should be false - and it is the case in Power Apps. The only self-contained Patch overload is the one that merges records (and that is not an asynchronous function)